### PR TITLE
Add RemoveNeverMatchingRuleFix

### PR DESCRIPTION
**Pull request by reacheight**

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/RemoveNeverMatchingRuleFix.fs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    override x.Text = "Remove never matching rule"
+    override x.IsAvailable _ = isValid matchClause
+
+    override x.ExecutePsiTransaction _ =
+        use writeLock = WriteLockCookie.Create(matchClause.IsPhysical())
+
+        if isLastChild matchClause then
+            let matchExpr = MatchExprNavigator.GetByClause(matchClause)
+            if isInlineSpaceOrComment matchExpr.NextSibling then
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/RemoveNeverMatchingRuleFix.fs]: `MatchExprNavigator.GetByClause` can return null. Consider the following:
```fsharp
function _ -> () | _ -> ()
```

File: ReSharper.FSharp/test/data/features/quickFixes/removeNeverMatchingRule/Multiple rules 01.fs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Module
+
+do
+    match () with
+    | _ -> ()
+    | _ -> (){caret}
+    | _ -> ()
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/quickFixes/removeNeverMatchingRule/Multiple rules 01.fs]: Nit pick: I'd probably use some named patterns to add some clarity to what is deleted and what is remained.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/RemoveNeverMatchingRuleFix.fs [hunk 3]
```diff
@@ -1,28 +1,39 @@
...
+open JetBrains.ReSharper.Psi.Tree
 open JetBrains.ReSharper.Resources.Shell
 
 type RemoveNeverMatchingRuleFix(warning: RuleNeverMatchedWarning) =
     inherit FSharpQuickFixBase()
 
     let matchClause = warning.MatchClause
 
+    let isNotMatchClause (node: ITreeNode) =
+        node.NodeType != ElementType.MATCH_CLAUSE.As<NodeType>()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/RemoveNeverMatchingRuleFix.fs]: `node :? IMatchClause` ? :)

**Comment by auduchinok** [comment]
Could you please also add a test for a case where you delete the first clause on a line, while preserving others, like the following?
```fsharp
let _ =
    match 0 with
    | 1 -> ()
    | 1 -> (){caret} | 2 -> ()
    | _ -> ()
```

**Comment by auduchinok** [comment]
@reacheight Thanks! It looks really great!

**Comment by NinoFloris** [comment]
What would happen with or patterns?

```
let _ =
    match 0 with
    | 1 -> ()
    | 1 | 2 -> ()
    | _ -> ()
```

**Comment by auduchinok** [comment]
@NinoFloris There would be no warning reported by FCS in that case, since these warnings are reported for whole match clauses and not individual patterns inside their left parts.

**Pull request merged**